From f181ae49d6fdea4eaabac7fb3c0639851017cceb Mon Sep 17 00:00:00 2001
From: Stivvo <stivvo01@gmail.com>
Date: Wed, 2 Sep 2020 17:37:14 +0200
Subject: [PATCH 1/3] Set pos and size with rules for floating windows

Rules support x, y, h, w values. They only take effect on floating
windows

When x == 0 || y = 0, the client is placed at the center of the monitor
When w == 0 || h = 0, the client default size is used
---
 config.def.h | 7 ++++---
 dwl.c        | 9 +++++++++
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/config.def.h b/config.def.h
index 53021cf74..15dbf056a 100644
--- a/config.def.h
+++ b/config.def.h
@@ -9,10 +9,11 @@ static const float focuscolor[]     = {1.0, 0.0, 0.0, 1.0};
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
 static const Rule rules[] = {
-	/* app_id     title       tags mask     isfloating   monitor */
+	/* app_id     title       tags mask     isfloating   monitor x  y  width heigh t*/
+	/* x, y, width, heigh = 0 -> use default */
 	/* examples:
-	{ "Gimp",     NULL,       0,            1,           -1 },
-	{ "firefox",  NULL,       1 << 8,       0,           -1 },
+	{ "Gimp",     NULL,       0,            1,           -1, 	0, 	 0,   500, 400 },
+	{ "firefox",  NULL,       1 << 8,       0,           -1, 	200, 100, 0,   0 },
 	*/
 };
 
diff --git a/dwl.c b/dwl.c
index 730e46a1f..67305e037 100644
--- a/dwl.c
+++ b/dwl.c
@@ -163,6 +163,10 @@ typedef struct {
 	unsigned int tags;
 	int isfloating;
 	int monitor;
+	int x;
+	int y;
+	int w;
+	int h;
 } Rule;
 
 /* Used to move all of the data necessary to render a surface from the top-level
@@ -357,6 +361,11 @@ applyrules(Client *c)
 			wl_list_for_each(m, &mons, link)
 				if (r->monitor == i++)
 					mon = m;
+			if (c->isfloating)
+				resize(c, r->x ? r->x + mon->w.x : mon->w.width / 2 - c->geom.width / 2 + mon->w.x,
+						r->y ? r->y + mon->w.y : mon->w.height / 2 - c->geom.height / 2 + mon->w.y,
+						r->w ? r->w : c->geom.width,
+						r->h ? r->h : c->geom.height, 1);
 		}
 	}
 	setmon(c, mon, newtags);

From f75641465094883cab5453cc40a9298f82891ba8 Mon Sep 17 00:00:00 2001
From: Stivvo <stivvo01@gmail.com>
Date: Wed, 2 Sep 2020 17:53:21 +0200
Subject: [PATCH 2/3] Instructions in def.h

---
 config.def.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 15dbf056a..73efb471a 100644
--- a/config.def.h
+++ b/config.def.h
@@ -10,11 +10,13 @@ static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
 static const Rule rules[] = {
 	/* app_id     title       tags mask     isfloating   monitor x  y  width heigh t*/
-	/* x, y, width, heigh = 0 -> use default */
 	/* examples:
 	{ "Gimp",     NULL,       0,            1,           -1, 	0, 	 0,   500, 400 },
 	{ "firefox",  NULL,       1 << 8,       0,           -1, 	200, 100, 0,   0 },
 	*/
+	/* x, y, width, heigh are floating only
+	* When x or y == 0 the client is placed at the center of the screen,
+	* when width or height == 0 the default size of the client is used*/
 };
 
 /* layout(s) */

From 7f0005c6efba1e616d78d36f97a4f8dbeb5c1172 Mon Sep 17 00:00:00 2001
From: Stivvo <stivvo01@gmail.com>
Date: Wed, 9 Sep 2020 18:45:17 +0200
Subject: [PATCH 3/3] Improve example comments in def.h

---
 config.def.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 73efb471a..8130f78a2 100644
--- a/config.def.h
+++ b/config.def.h
@@ -9,7 +9,7 @@ static const float focuscolor[]     = {1.0, 0.0, 0.0, 1.0};
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
 static const Rule rules[] = {
-	/* app_id     title       tags mask     isfloating   monitor x  y  width heigh t*/
+	/* app_id     title       tags mask   isfloating    monitor x    y   width height */
 	/* examples:
 	{ "Gimp",     NULL,       0,            1,           -1, 	0, 	 0,   500, 400 },
 	{ "firefox",  NULL,       1 << 8,       0,           -1, 	200, 100, 0,   0 },
